### PR TITLE
fast shared memory communication and other improvements

### PR DESCRIPTION
**Pull request by mikeroberts3000**
- fast cross-platform shared memory communication, leading to a 2x end-to-end speedup for large image observations on all platforms
- SimulationController still uses RPC communication by default, but Tasks/Agents/etc can can opt into fast shared memory communication on a case-by-case basis by setting a flag when returning a Box object.
- introduced boost as a dependency, removed asio as a dependency because it is part of boost
- found uses for boost throughout the codebase
- improved ASSERT behavior, will now wait for keyboard input and can optionally break into the debugger
- improved debugging experience, CoreUtils optionally waits for keyboard input during initialization to attach a debugger to the running executable, which makes it easy to debug the executable in the exact state it is in when launching from Python
- improved debugging experience, fixed broken breakpoint functionality on XCode and LLDB due do our use of symbolic links
- other code cleanup tasks

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,5 +1,7 @@
 __pycache__
 
+cpp/.vs
```
Comment (RachithP) [inline .gitignore:3]: Don't need this as it's system specific config.

File: cpp/unreal_projects/SpearSim/Source/SpearSim/SpearSim.Build.cs [hunk 2]
```diff
@@ -9,15 +9,17 @@ public class SpearSim : ModuleRules
...
         PrivatePCHHeaderFile = "";
         bUseUnity = false;
 
-        // Turn off code optimization except in shipping builds for faster build times
+        // Turn off code optimization except in shipping builds for faster build times.
         OptimizeCode = ModuleRules.CodeOptimization.InShippingBuildsOnly;
 
-        // Enable exceptions because some of our third-party dependencies use them
+        // Our ASSERT macro throws exceptions, and so does our templated function Config::get(...),
+        // because it depends yaml-cpp, which throws exceptions. So we need to enable exceptions
```
Comment (RachithP) [inline cpp/unreal_projects/SpearSim/Source/SpearSim/SpearSim.Build.cs:21]: `it depends yaml-cpp` -> `it depends on yaml-cpp`

File: cpp/unreal_plugins/OpenBot/Source/OpenBot/OpenBot.Build.cs [hunk 3]
```diff
@@ -11,15 +11,17 @@ public class OpenBot : ModuleRules
...
         PrivatePCHHeaderFile = "";
         bUseUnity = false;
 
-        // Turn off code optimization except in shipping builds for faster build times
+        // Turn off code optimization except in shipping builds for faster build times.
         OptimizeCode = ModuleRules.CodeOptimization.InShippingBuildsOnly;
 
-        // Enable exceptions because some of our third-party dependencies use them
+        // Our ASSERT macro throws exceptions, and so does our templated function Config::get(...),
+        // because it depends yaml-cpp, which throws exceptions. So we need to enable exceptions
```
Comment (RachithP) [inline cpp/unreal_plugins/OpenBot/Source/OpenBot/OpenBot.Build.cs:23]: 'it depends yaml-cpp' -> 'it depends `on` yaml-cpp'

File: cpp/unreal_plugins/CoreUtils/Source/CoreUtils/Config.h [hunk 4]
```diff
@@ -36,16 +36,7 @@ class COREUTILS_API Config
...
-        std::stringstream ss(key);
-        std::string s;
-        std::vector<std::string> keys;
-
-        while (std::getline(ss, s, '.')) {
-            keys.push_back(s);
-        }
-
-        return getFromKeys<TValue>(keys);
+        return getFromKeys<TValue>(Std::tokenize(key, "."));
```
Comment (RachithP) [inline cpp/unreal_plugins/CoreUtils/Source/CoreUtils/Config.h:39]: Good usage of boost.

**Review by RachithP — APPROVED** [review]
Some minor comments. 

Looks good to me!

File: .gitignore (hunk 1 above)
Comment (mikeroberts3000) [reply .gitignore:3]: I don't mind having this here. Yes, it is system-specific, but at the same time, I'd like for Visual Studio to be an "officially supported" way of debugging on Windows.

At the same time, I don't want to maintain IDE projects for every platform, but I'm ok with ignoring vs temp files.

File: cpp/unreal_projects/SpearSim/Source/SpearSim/SpearSim.Build.cs (hunk 2 above)
Comment (mikeroberts3000) [reply cpp/unreal_projects/SpearSim/Source/SpearSim/SpearSim.Build.cs:21]: Good catch.

File: cpp/unreal_plugins/OpenBot/Source/OpenBot/OpenBot.Build.cs (hunk 3 above)
Comment (mikeroberts3000) [reply cpp/unreal_plugins/OpenBot/Source/OpenBot/OpenBot.Build.cs:23]: Good catch.

File: cpp/unreal_plugins/CoreUtils/Source/CoreUtils/Config.h (hunk 4 above)
Comment (mikeroberts3000) [reply cpp/unreal_plugins/CoreUtils/Source/CoreUtils/Config.h:39]: Agreed.

File: python/spear/config/default_config.simulation_controller.yaml [hunk 5]
```diff
@@ -154,18 +154,22 @@ SIMULATION_CONTROLLER:
...
+    # those values in the engine. When Unreal defines UIMin/UIMax vales, it will
+    # only clamp those values values in the editor.
     SCENE_CAPTURE_COMPONENT_CAPTURE_SOURCE: "SCS_FinalToneCurveHDR"     # "SCS_FinalColorHDR", "SCS_FinalToneCurveHDR"
     FINAL_COLOR_OVERRIDE_AUTO_EXPOSURE_SPEED_UP: True
     FINAL_COLOR_AUTO_EXPOSURE_SPEED_UP: 9999.0                          # ClampMin=0.02, UIMax=20.0
     FINAL_COLOR_OVERRIDE_AUTO_EXPOSURE_SPEED_DOWN: True
     FINAL_COLOR_AUTO_EXPOSURE_SPEED_DOWN: 9999.0                        # ClampMin=0.02, UIMax=20.0
     FINAL_COLOR_OVERRIDE_INDIRECT_LIGHTING_INTENSITY: True
-    FINAL_COLOR_INDIRECT_LIGHTING_INTENSITY: 0.0                        # set to 0.0 on Windows for ray-traced global illumination, set to 1.0 on macOS and Linux to use baked lighting
+    FINAL_COLOR_INDIRECT_LIGHTING_INTENSITY: 0.0                        # # set to 0.0 for raytracing, set to 1.0 for baked
```
[the remark this answers is not in the document]
Comment (mikeroberts3000) [reply python/spear/config/default_config.simulation_controller.yaml]: Good catch.

**Pull request merged**